### PR TITLE
Add invites functionality

### LINKED ISSUE #75: Removing the allowlist and replacing with admin-generated invite codes
the allowlist functionality originally existed to populate the merveilles forum using pre-authorized domains where merv-adjacent people could be found. this was in the form of 1) the merveilles town mastodon instance, and 2) [xxiivv webring](https://webring.xxiivv.com/) denizens.

users of the mastodon instance were encouraged to temporarily put the signup-generated redemption code into their profile information, which would then be scraped. however, mastodon has gone away from providing static html in preference of dynamic, javascript rendered html rendering option 1) obsolete. as the forum is mostly populated now, and there is the possibility to request invites from known admin users, i don't feel like the allowlist functionality has relevancy anymore

wdy @decentral1se, will [pmc](https://permacomputing.net/) use this functionality?

---

a more modular replacement could instead be:

1) admins can create invite codes. an invite code is a static string which can be redeemed by users in the registration flow. an invite code may have a pre-specified number of uses.
2) pinned threads.

combining these two we can achieve the following: a private and pinned thread that is posted by an admin and, by way of edits, contains the latest invite code could act as a simple member-based invite system.

thoughts welcome :]

### PR DESCRIPTION
**Pull request by cblgh**
This feature adds invite functionality to the forum. An invite allows a person without an account to redeem the invite during account registration, granting entry to the forum as a member.

cerca-invites-wip

Invites are created by admins and they are UUID v4 identifiers. Invites have yet to be claimed are displayed as either a plain UUID v4 string or formatted as a link to the forum's `/register` route, with the invite as a URL parameter.

If a forum community wants to let its users freely invite their friends, admins can, for example, generate a large batch of invites and post them in a private thread.

Invites can optionally be labeled, to allow for creating different batches for different purposes and keep the batches separate so that they are not "redeemed too quickly".
 
closes  #75

### CODE REVIEW

**Comment by cblgh** [comment]
a thought:  

it  would be nice for part of this functionality to be callable by a script e.g. by invoking a standalone cmd. it would enable having a service that can act as an intermediary for creating & returning invites. the functionality could be called by a IM bot or be part of a service with an allowlist. like the original allowlist functionality of the forum except more flexible and mostly living outside of this repo :)

**Comment by decentral1se** [comment]
Incredible, just driving by with a quick read, haven't taken it for a test ride. It looks very complete so far!

One thing that comes to mind: is it possible to set an invite as redeemable but then not expiring (unless you manually click expire) for the "open invite only" mode. This would make it more convenient than having to nag the admin to generate a big batch of invites?

**Comment by cblgh** [comment]
> One thing that comes to mind: is it possible to set an invite as redeemable but then not expiring (unless you manually click expire) for the "open invite only" mode. This would make it more convenient than having to nag the admin to generate a big batch of invites?

so for instance creating one invite and when it is redeemed, not removing it from the database - is this kind of what you are asking for?

**Comment by decentral1se** [comment]
> so for instance creating one invite and when it is redeemed, not removing it from the database - is this kind of what you are asking for?

yep, it can be redeemed endlessly and never expire until you delete it.

this could avoid busy work on the long-term for non-admin people wanting to self-organise invites without having to ask for new invite batches each time.

**Comment by cblgh** [comment]
> yep, it can be redeemed endlessly and never expire until you delete it.
> 
> this could avoid busy work on the long-term for non-admin people wanting to self-organise invites without having to ask for new invite batches each time.

i can see how that would be useful! could be pinned in a local-only post as well. still chewing it over :]

the flip-side: it could also e.g. be used to spam an abandoned forum if a register link with a "perpetual invite code" was posted openly. in some sense this possibility brings to mind the [semi-recent fedi / masto kerfuffle with open registration](https://mastodon.social/@Gargron/111953045633249137), where a default on abandoned servers was exploited by spammers

if i do bring it in, i'll consider having it behind a config toggle

File: database/moderation.go [hunk 1]
```diff
@@ -571,14 +574,15 @@ func (d DB) ClaimInvite (invite string) (bool, error) {
...
+    BatchQuery{desc: "get invite code's batchid", stmt:"SELECT batchid FROM invites WHERE invite = ?"},
     BatchQuery{desc: "delete invite from table", stmt:"DELETE FROM invites WHERE invite = ?"},
   }
 
-  for _, operation := range ops {
-    operation.preparedStmt, err = tx.Prepare(operation.stmt)
-    defer operation.preparedStmt.Close()
+  for i, operation := range ops {
+    ops[i].preparedStmt, err = tx.Prepare(operation.stmt)
+    defer ops[i].preparedStmt.Close()
```
Comment (cblgh) [inline database/moderation.go:583]: classic golang. i think i came up with a good mnemonic to reduce errors in this department.

if ranging over e.g. a slice and wanting to change something, never use the **second** parameter of range.

```golang
for i, _ := range items { items[i].member = ... }
```

while, if ranging over e.g. a slice and only wanting to consume items, never use the **first** parameter of range:



```golang
for _, item := range items { use(item) }
```

**Comment by cblgh** [comment]
@decentral1se give 'er a spin whenever you have time! i'm in no rush to merge it in, letting the decision on whether to include https://github.com/cblgh/cerca/tree/reusable-invites-behind-flag simmer. 

from my pov it's complete wrt what i set out to do, so anything that's not thought of is an oversight on my part and i'd love to know abt it :~)

**Comment by decentral1se** [comment]
@cblgh flawless victory! works *so* well, congrats 🏳️‍🌈

the only other thing that came to mind is a "counter" for the reusable invite, which can be a visual reminder to admins to see who many times it is used. if it is suddenly goes from 1 to 300 then they can also disable it quickly.

in general, im fine with leaving it out if you like! or doing it behind a config flag, once the other ticket gets rounded off. its deffo for a future busier cerca which i currently don't need to personally deal with.

nice one!

**Comment by cblgh** [comment]
thanks for having a look ^__________^

> the only other thing that came to mind is a "counter" for the reusable invite, which can be a visual reminder to admins to see who many times it is used. if it is suddenly goes from 1 to 300 then they can also disable it quickly.

glad you raised this! it was on my mind too. the code currently makes a note of the registration origin, which previously was just a domain and reg link and is now the invite code that was used. i could do a simple tally with a sql query and toss it into the admin view under a `<details>` element?

**Comment by decentral1se** [comment]
that sounds extremely legit! also a bonus i guess but if you're keen to get it in, i'm keen to test it. otherwise, hit the green button

File: database/moderation.go [hunk 2]
```diff
@@ -511,3 +530,255 @@ func (d DB) GetAdmins() []User {
...
+* iterate over the map
+*		-> collect creation time
+*   -> sort by time, newest first?
+*/
+
+// TODO (2024-11-22): consider d1's request of invites that are repeat-redeemable until deleted by an admin
+// returns: 
+// * a bool signaling whether the invite was successfully claimed or not
+// * a string representing the batchid associated with this invite
+// * an error, nil if everything went according to expectations (either claimed invite, or invite code was not valid)
```
Comment (cblgh) [inline database/moderation.go:556]:
```suggestion
````

File: database/moderation.go [hunk 3]
```diff
@@ -511,3 +530,255 @@ func (d DB) GetAdmins() []User {
...
+// CREATE TABLE IF NOT EXISTS invites (
+// id INTEGER PRIMARY KEY AUTOINCREMENT,
+// batchid TEXT NOT NULL, -- this is a uuid v4
+// invite TEXT NOT NULL,
+// label TEXT,
+// adminid INTEGER NOT NULL,
+// time DATE NOT NULL,
+//
+// FOREIGN KEY(adminid) REFERENCES users(id)
+// );
```
Comment (cblgh) [inline database/moderation.go:637]:
```suggestion
```

File: database/moderation.go [hunk 4]
```diff
@@ -511,3 +530,255 @@ func (d DB) GetAdmins() []User {
...
+	ed.Check(err, "prep stmt")
+	defer stmt.Close()
+
+	_, err = stmt.Exec(batchid)
+	util.Check(err, "execute delete")
+}
+
+/* TODO (2024-12-01): next up - start to write these database routines to parts that called from the server route
+* handlers :) */
+
```
Comment (cblgh) [inline database/moderation.go:721]:
```suggestion
```

File: server/server.go [hunk 5]
```diff
@@ -667,66 +650,37 @@ func (h RequestHandler) RegisterRoute(res http.ResponseWriter, req *http.Request
...
-				return
-			}
-			// check verification link domain against allowlist
-			if !util.Contains(h.allowlist, u.Host) {
-				fmt.Println(h.allowlist, u.Host, util.Contains(h.allowlist, u.Host))
-				renderErr("Verification link's host (%s) is not in the allowlist", u.Host)
-				return
-			}
+		// // skip invite code during dev registering
+		// if !developing {
```
Comment (cblgh) [inline server/server.go:679]: decide on whether to have this dev functionality or nae

**Comment by decentral1se** [comment]
> ait ait here we go wdyt

Oh yes @cblgh, that's excellent 🎉

**Comment by cblgh** [comment]
m'kay! i'll sprinkle a little more fairy dust on this and then merge this weekend?! :>:>

File: README.md [hunk 6]
```diff
@@ -64,7 +62,7 @@ cerca adduser --username "<username>"
 Cerca supports community customization.
 
 * Write a custom [about text](/defaults/sample-about.md) describing the community inhabiting the forum
-* Define your own [registration rules](/defaults/sample-rules.md), [how to verify one's account](/defaults/sample-verification-instructions.md), and link to an existing code of conduct
+* Define your own [registration rules](/defaults/sample-rules.md), [instructions on getting an invite code to register](/defaults/sample-registration-instructions.md), and link to an existing code of conduct
 * Set your own [custom logo](/defaults/sample-logo.html) (whether svg, png or emoji)
 * Create your own theme by writing plain, frameworkless [css](/html/assets/theme.css)
```
Comment (cblgh) [inline README.md:67]: add to feature section: 
* **Invites**: Fully-featured system for creating invites, both one-time usable invites and reusable invites. Admins can monitor invite redemption by batch as well as issue and delete batches of invites. Accessible using the same simple type of web interface that services the rest of the forum's administration tasks.

**Comment by cblgh** [comment]
little loopback on the usage of this feature:

even for our already populated [forum](https://forum.merveilles.town), 31 new people signed up to make an account! what a great suggestion for the multi-use invite and the addition of a counter for visibility @decentral1se :)

**Pull request merged**